agents/timestream_aggregator/timestream_aggregator.py

```diff
@@ -114,40 +410,19 @@ def start_aggregation(self, session, params=None):
 
         self.log.info("Data directory set to {}".format(self.data_dir))
         self.log.info("New file every {} seconds".format(self.time_per_file))
-        self.log.info("Listening to {}:{}".format(self.address, self.port))
+        self.log.info("Listening to {}".format(self.address))
 
-        reader = core.G3Reader("tcp://{}:{}".format(self.address,
-                                                    self.port))
-        writer = None
-
-        last_meta = None
         self.is_streaming = True
 
+        recorder = FrameRecorder(self.time_per_file, self.address,
+                                 self.data_dir, self.log)
+
         while self.is_streaming:
-            if writer is None:
-                start_time = time.time()
-                filepath = _create_file_path(start_time, self.data_dir)
-                self.log.info("Writing to file {}".format(filepath))
-                writer = core.G3Writer(filename=filepath)
-
-                if last_meta is not None:
-                    writer(last_meta)
-
-            frames = reader.Process(None)
-            for f in frames:
-                if f.type == core.G3FrameType.Observation:
-                    last_meta = f
-                writer(f)
-                writer.Flush()
-
-            if (time.time() - start_time) > self.time_per_file:
-                writer(core.G3Frame(core.G3FrameType.EndProcessing))
-                writer = None
+            recorder.run()
```

Comment:
I think this loop won't always exit in a timely fashion when self.is_streaming is set False.  Consider the case where no G3NetworkSender is present -- the read_frames() function will block... because _establish_reader_connection() will block.

And _establish_reader_connection() can block in the FrameRecorder constructor, which is equally problematic..

I think the solution is to remove the _establish... call from the constructor -- it will get called on first run(), anyway.  Then rewrite the _establish_reader_connection() and run() functions a little so that run() will always return within 1 second or so (probably 5 seconds sometimes...), even if all it has done is to fail to make a connection.

Reply:
Hmm, right. I'm going to have to give this some thought to avoid breaking desired functionality in other scenarios. If I remove the _establish call in the constructor then the first `self.reader.Process(None)` call in `read_frames()` won't be available. If I establish first in that call then I'd be reestablishing the connection every time I read, which I don't think we want.

Even once we make that non-blocking then `create_new_file` would start making new files even if frames weren't coming in (we were relying on `read_frames()` to block until frames came in to prevent this.) I think most of the other methods need adjustment then too to check if reader is instantiated. Alternatively I suppose the check could happen in run(), where we skip these other things if we were unable to `read_frames()`.

Reply:
This is kind of why I like having the frame reading in a separate thread. If you want to do something like the HKAggregator, you could start a thread that continuously reads frames to a queue and tries to re-establish connections while streaming is true (just make sure that there's no way it can block forever).

Then in the run function you can do something like:
```
if self.frames:
    if writer is not None:
        create new file
    write frames to file

if (time.time() - last_write_time > file_limit):
    close file and set writer to None
```

Reply:
The threads idea is interesting, if you want to go that route.  But that's extra, because you still need to handle doing a clean exit from run() in cases that you can't make a reader or you can't get frames.  (Unless you use Exceptions.)

I propose this minimal fix:
- _establish_reader_connection() tries only once and then returns.
- read_frames() should check self.reader is None, and try (once) to _establish_reader_connection().  If that fails, it returns.  Otherwise, read_frames() should read some frames.  But if that fails, it should clean up (as it does now) and return.
- In run(), after read_frames(), check if self.frames.  If not, return immediately instead of calling the other 4 functions.

Also, two other things:
- Currently you avoid rapid reconnection attempts by sleep(1) in the _establish_reader_connection loop.  If not looping any more, one way to limit the rate of reconnections is to compare the present time to some stored "last_reader_connect_time", and then sleep a bit if that time was too recent.
- In close_file(), the log message "File closed" is either potentially misleading should be in the "if" block.

Reply:
Thanks, this was helpful, but now I run into the situation where I need to distinguish between a timeout and the streamer being off.

After making the suggested changes I realized that if `read_frames()` checks for `self.reader is None` to reestablish the connection, then we'll want to set `self.reader = None` if the connection times out, so I started to do this if no frames came in. However, this would then cause the connection to break and reestablish if the streamer isn't gone, but simply not sending frames. After several of these iterations the aggregator locks up on a connection reestablishment (is there some limit to the number of connections a G3NetworkSender will make?)

I got caught up prepping a cooldown here, so got to this later than I thought. I'll get back to it in the morning.

Reply:
Do we have to distinguish them? I think that once we have the ability to detect when rogue has actually stopped streaming, and close the connection on the NetworkSender side, we won't need the timeout functionality at all.

Reply:
I have looked into this a bit today... I think it's a real problem.  The G3NetworkSender currently only handles the accepting of new connections, and the termination of closed connections, in Process().  So if no frames are being written, the connection requests (or dead connections) just pile up.

I looked into a few approaches to this but I think the only quick fix is to have the smurf-streamer send, on ~1 second intervals, some kind of keep-alive frame (that will be simply discarded in stream-processor).  The structure of these frames should fit into a general schema for "flow control" frames that will also be used to announce start / stop of data acquisition.

The potential problem with simply closing the Sender down whenever there's no frame data to process is that it means a whole bunch of things need to get set up in the short time after the first couple of frames are received from Rogue.  The Sender needs to set up the server while also queuing frames; the Reader needs to connect quickly enough that those queued frames don't start falling on the floor.  That kind of rush is asking for trouble.

Reply:
Alright, all aspects of your minimal fix have been addressed @mhasself. Some additional changes were needed once those were made, as we've discussed in Slack. Details are in 
aa1e28a.

We also now handle control frames, but currently only keep alive frames, not start/end frames.